rust/src/storage/file.rs

```diff
@@ -58,26 +65,110 @@ impl StorageBackend for FileStorageBackend {
     }
 
     async fn put_obj(&self, path: &str, obj_bytes: &[u8]) -> Result<(), StorageError> {
-        let dir = std::path::Path::new(path);
-
-        if let Some(d) = dir.parent() {
-            fs::create_dir_all(d).await?;
-        }
+        let tmp_file = self
+            .root
+            .join("_tmp_writes")
+            .join(Uuid::new_v4().to_string());
 
-        // use `OpenOptions` with `create_new` to create the file handle.
-        // this will force ErrorKind::AlreadyExists if the file already exists.
-        let mut f = fs::OpenOptions::new()
-            .create_new(true)
-            .write(true)
-            .open(path)
-            .await
-            .map_err(|e| match e.kind() {
-                std::io::ErrorKind::AlreadyExists => StorageError::AlreadyExists(path.to_string()),
-                _ => StorageError::Io { source: e },
-            })?;
+        create_parent_dirs(&tmp_file).await?;
+        create_parent_dirs(Path::new(path)).await?;
 
+        let mut f = fs::File::create(&tmp_file).await?;
         f.write(obj_bytes).await?;
 
+        rename(tmp_file.to_str().unwrap(), path)?;
```

Comment:
`to_str` on path returns Option, leave it as is with `unwrap` or there's fancy workaround here?

Reply:
You can avoid this unwrap by changing signature of rename to accept `AsRef<Path>`, basically something like [this Rust playground example](https://play.rust-lang.org/?code=%23!%5Ballow(unused)%5D%0Afn%20main()%20%7B%0Ause%20std%3A%3Apath%3A%3A%7BPath%2C%20PathBuf%7D%3B%0A%0Alet%20path%20%3D%20Path%3A%3Anew(%22%2Ftmp%2Ffoo.txt%22)%3B%0Aassert_eq!(path.with_file_name(%22bar.txt%22)%2C%20PathBuf%3A%3Afrom(%22%2Ftmp%2Fbar.txt%22))%3B%0A%0Alet%20path%20%3D%20Path%3A%3Anew(%22%2Ftmp%22)%3B%0Aassert_eq!(path.with_file_name(%22var%22)%2C%20PathBuf%3A%3Afrom(%22%2Fvar%22))%3B%0A%7D&edition=2018).

Then inside rename function convert Path to OsString, then to CString.

Reply:
Done that in https://github.com/delta-io/delta-rs/pull/126/commits/44805f34b88b09a466b5502ee295a4aa184f4044#diff-eed9f66f54ec53b6ccce2167c4fb49e27e075536e5d46390dfcd3fc92da18d2bL112, but still needed some map_err within rename, which was better. 

However I found that there's `join_path` which works with `String`s, and I needed `Path` only for joining. So basically reverted all changes for `Path`/`PathBuf` and using plain string sinstead.